### PR TITLE
internal(backend|curve): use regexp for platform

### PR DESCRIPTION
**Pull request by elais**
The previous implementation required the configuration to use the exact
device name (i.e corsaircpro-hid-3-7 or amdgpu-pci-0d00) which meant if
a device booted to a different hwmon path (causing corsaircpro-hid-3-7
to be corsaircpro-hid-3-6) then it would not get picked up by fan2go.

This makes fan2go more powerful and immune to underlying changes to the
device interface that may happen when booting into a system. It also  it low key makes the README more accurate :-).

### CODE REVIEW

**Comment by markusressel** [comment]
Hmm I really hope multiple devices on the same usb controller can still reliably be specified though, otherwise the whole reason why I started fan2go in the first place would not be fulfilled.
I wanted to match the output of `sensors` as closely as possible, seems like that might not be the best idea regarding naming consistency 🤔

File: cmd/curve.go [hunk 1]
```diff
@@ -29,7 +30,11 @@ var curveCmd = &cobra.Command{
 		for _, config := range configuration.CurrentConfig.Fans {
 			if config.HwMon != nil {
 				for _, controller := range controllers {
-					if controller.Platform == config.HwMon.Platform {
+					matched, err := regexp.MatchString(config.HwMon.Platform, controller.Platform)
```
Comment (markusressel) [inline cmd/curve.go:33]: If we change the behaviour again anyway, can we introduced a "case-insensitive" search for all of the comparisons too?
I had problems in the past with that too.

File: cmd/curve.go [hunk 2]
```diff
@@ -29,7 +30,11 @@ var curveCmd = &cobra.Command{
...
+					if err != nil {
+						ui.Fatal("Unable to read curve platform: %s", config.HwMon.Platform)
```
Comment (markusressel) [inline cmd/curve.go:35]: Can we make this error message (and the others) more specific and add context please?
Like: `"Failed to match platform regex of %s (%s) against controller platform %s", config.ID, config.HwMon.Platform, controller.Platform`

File: internal/backend.go [hunk 3]
```diff
@@ -128,7 +129,11 @@ func InitializeObjects() {
 		if config.HwMon != nil {
 			found := false
 			for _, c := range controllers {
-				if c.Platform == config.HwMon.Platform {
+				matched, err := regexp.MatchString(config.HwMon.Platform, c.Platform)
+				if err != nil {
+					ui.Fatal("Unable to read sensor platform: %s", config.HwMon.Platform)
```
Comment (markusressel) [inline internal/backend.go:134]: Same thing here

File: internal/backend.go [hunk 4]
```diff
@@ -128,7 +129,11 @@ func InitializeObjects() {
 		if config.HwMon != nil {
 			found := false
 			for _, c := range controllers {
-				if c.Platform == config.HwMon.Platform {
+				matched, err := regexp.MatchString(config.HwMon.Platform, c.Platform)
```
Comment (markusressel) [inline internal/backend.go:132]: Same here

File: internal/backend.go [hunk 5]
```diff
@@ -174,7 +179,11 @@ func InitializeObjects() {
...
```
Comment (markusressel) [inline internal/backend.go:182]: Same here

File: internal/backend.go [hunk 6]
```diff
@@ -174,7 +179,11 @@ func InitializeObjects() {
...
+				if err != nil {
+					ui.Fatal("Unable to read fan platform: %s", config.HwMon.Platform)
```
Comment (markusressel) [inline internal/backend.go:184]: Same here

**Review by markusressel — CHANGES_REQUESTED** [review]

**Comment by elais** [comment]
> Hmm I really hope multiple devices on the same usb controller can still reliably be specified though, otherwise the whole reason why I started fan2go in the first place would not be fulfilled. I wanted to match the output of `sensors` as closely as possible, seems like that might not be the best idea regarding naming consistency thinking

I think it will be fine. For instance if you run this with your config it shouldn't require any changes to continue doing what it's doing.

> I really hope multiple devices on the same usb controller can still reliably be specified though, otherwise the whole reason why I started fan2go in the first place would not be fulfilled.

Here's my current fan2go sensors config

```yaml
# A list of sensors to monitor
sensors:
  # A user defined ID, which is used to reference
  # a sensor in a curve configuration (see below)
  - id: water_temp
    # The type of sensor configuration
    hwmon:
      # The controller platform as displayed by `fan2go detect`, f.ex.:
      # "nouveau", "coretemp" or "it8620" etc.
      platform: corsaircpro
      # The index of this sensor as displayed by `fan2go detect`
      index: 1
  - id: ambient_temp_1
    # The type of sensor configuration
    hwmon:
      # The controller platform as displayed by `fan2go detect`, f.ex.:
      # "nouveau", "coretemp" or "it8620" etc.
      platform: corsaircpro
      # The index of this sensor as displayed by `fan2go detect`
      index: 2

  - id: ambient_temp_2
    # The type of sensor configuration
    hwmon:
      # The controller platform as displayed by `fan2go detect`, f.ex.:
      # "nouveau", "coretemp" or "it8620" etc.
      platform: corsaircpro
      # The index of this sensor as displayed by `fan2go detect`
      index: 3

  - id: ambient_temp_3
    # The type of sensor configuration
    hwmon:
      # The controller platform as displayed by `fan2go detect`, f.ex.:
      # "nouveau", "coretemp" or "it8620" etc.
      platform: corsaircpro
      # The index of this sensor as displayed by `fan2go detect`
      index: 4
```

I'm reading four sensors off of one device that I'm matching using regexp (running fan2go from the project folder). For most users it should be a matter of a more fine tuned regexp to find the correct controller. Regexp is powerful enough to find the correct controller and if a user has doubts, they can always just use the full name. HID devices are fickle though, changing the numbers at the end and making things that should find them immediately run the risk of failing on reboot. I think it will be fine but if you can think of any edge cases that need extra considering I'm willing to add a test case for them.

File: cmd/curve.go (hunk 1 above)
Comment (elais) [reply cmd/curve.go]: Made the search case insenitive

File: internal/backend.go (hunk 4 above)
Comment (elais) [reply internal/backend.go]: Added the case insenitive search

File: README.md [hunk 7]
```diff
@@ -129,8 +129,8 @@ sensors:
   - id: cpu_package
     # The type of sensor configuration, one of: hwmon | file
     hwmon:
-      # The controller platform as displayed by `fan2go detect`, f.ex.:
-      # "nouveau", "coretemp" or "it8620" etc.
+      # A regex matching a controller platform displayed by `fan2go detect`, f.ex.:
+      # "coretemp", "it8620", "corsaircpro-*" etc.
```
Comment (markusressel) [inline README.md:133]: Thinking about it, this should be `.*` at the end, right? So:
`corsaircpro-.*`

Or can `regexp.MatchString` deal with the `*` at the end?

**Review by markusressel — CHANGES_REQUESTED** [review]

**Comment by markusressel** [comment]
> I'm reading four sensors off of one device that I'm matching using regexp (running fan2go from the project folder). For most users it should be a matter of a more fine tuned regexp to find the correct controller. Regexp is powerful enough to find the correct controller and if a user has doubts, they can always just use the full name. HID devices are fickle though, changing the numbers at the end and making things that should find them immediately run the risk of failing on reboot. I think it will be fine but if you can think of any edge cases that need extra considering I'm willing to add a test case for them.

No I think its fine. We will just try it out and see if there are still edge cases lateron.
Thx again for your contribution! :heart:

File: README.md (hunk 7 above)
Comment (elais) [reply README.md:133]: `regexp.MatchString` can deal with the `*` at the end. When matching I believe its matching for the regexp inside of the string which is fine enough for our purposes. I just ran a quick test using my own config.
